### PR TITLE
[Actions + Triggers] Squarespace #2944

### PR DESCRIPTION
**Pull request by lcaresia**

### CODE REVIEW

**Comment by dylburger** [comment]
Thanks for submitting this PR! When we review PRs, we follow the [Pipedream component guidelines](https://pipedream.com/docs/components/guidelines/). If you're not familiar, here's a quick checklist:

- [ ] [Create components to address specific use cases whenever possible](https://pipedream.com/docs/components/guidelines/#component-scope)
- [ ] [Component `key`s should follow the format `app_name_slug-slugified-component-name`](https://pipedream.com/docs/components/guidelines/#component-key-pattern)
- [ ] Components should follow the [standard directory structure](https://pipedream.com/docs/components/guidelines/#folder-structure)
- [ ] [Prefer Node.js client libraries to REST APIs](https://pipedream.com/docs/components/guidelines/#using-apis-vs-client-libraries)
- [ ] When making API requests, [handle pagination](https://pipedream.com/docs/components/guidelines/#pagination) to ensure all data / events are processed
- [ ] [Use `secret` props](https://pipedream.com/docs/components/guidelines/#capturing-sensitive-data) to capture sensitive data
- [ ] Props and methods should be defined in [app files](https://pipedream.com/docs/components/guidelines/#app-files) whenever possible
- [ ] Document methods with [JS Docs](https://pipedream.com/docs/components/guidelines/#methods)
- [ ] [Use `optional` props](https://pipedream.com/docs/components/guidelines/#optional-vs-required-props) whenever possible, and [set a `default` value](https://pipedream.com/docs/components/guidelines/#default-values) where you can
- [ ] Use [async options](https://pipedream.com/docs/components/guidelines/#async-options) to accept user input wherever possible
- [ ] [Create subscription webhooks sources (vs polling sources) whenever possible](https://pipedream.com/docs/components/guidelines/#webhook-vs-polling-sources)
- [ ] [Always emit a `summary` with events](https://pipedream.com/docs/components/guidelines/#emit-a-summary)
- [ ] [Use a built-in deduping strategy wherever possible](https://pipedream.com/docs/components/guidelines/#deduping)
- [ ] When building webhook sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#webhook-sources)
- [ ] When building polling sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#polling-sources)

File: components/squarespace/actions/create-product/create-product.mjs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    storePageId: {
+      propDefinition: [
+        squarespace,
+        "storePageId",
+      ],
+    },
+    variants: {
+      label: "Variants",
+      description: "The variants of the product. Array must contain at least one object but no more than 100 objects. E.g `[ { \"sku\": \"SQ0557856\", \"pricing\": { \"basePrice\": { \"currency\": \"USD\", \"value\": \"12.95\" }, \"onSale\": false, \"salePrice\": { \"currency\": \"USD\", \"value\": \"0.00\" } }, \"stock\": { \"quantity\": 10, \"unlimited\": false }, \"attributes\": { \"Flavor\": \"Habanero\" }, \"shippingMeasurements\": { \"weight\": { \"unit\": \"POUND\", \"value\": 2.0 }, \"dimensions\": { \"unit\": \"INCH\", \"length\": 7.0, \"width\": 5.0, \"height\": 5.0 } } } ]`",
+      type: "string",
```
Comment (feyzullah) [inline components/squarespace/actions/create-product/create-product.mjs:21]:
```suggestion
      type: "string[]",
```
Optionally how about a string array here?

File: components/squarespace/actions/get-product/get-product.mjs [hunk 2]
```diff
@@ -0,0 +1,29 @@
+import squarespace from "../../squarespace.app.mjs";
+
+export default {
+  key: "squarespace-get-product",
+  name: "Get Product",
+  description: "Get a specific product. [See docs here](https://developers.squarespace.com/commerce-apis/create-product)",
```
Comment (feyzullah) [inline components/squarespace/actions/get-product/get-product.mjs:6]: I guess the doc link is incorrect

File: components/squarespace/sources/new-create-product/new-create-product.mjs [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    squarespace,
+    timer: {
+      type: "$.interface.timer",
+      default: {
+        intervalSeconds: 60 * 5,
+      },
+    },
+  },
+  async run() {
+    const products = await this.squarespace.getProducts();
```
Comment (feyzullah) [inline components/squarespace/sources/new-create-product/new-create-product.mjs:20]: I guess we cannot use webhooks here, right? While employing polling, is not `this.squarespace.getProducts` retrieving all products? Please correct me if I am wrong. If it retrieves all products, can we add some logic to filter new products?

File: components/squarespace/sources/new-create-transaction/new-create-transaction.mjs [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    squarespace,
+    timer: {
+      type: "$.interface.timer",
+      default: {
+        intervalSeconds: 60 * 5,
+      },
+    },
+  },
+  async run() {
+    const transactions = await this.squarespace.getTransactions();
```
Comment (feyzullah) [inline components/squarespace/sources/new-create-transaction/new-create-transaction.mjs:20]: As in `new-create-product`, is it returning all transactions or new transactions?

File: components/squarespace/sources/new-update-product/new-update-product.mjs [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    squarespace,
+    timer: {
+      type: "$.interface.timer",
+      default: {
+        intervalSeconds: 60 * 5,
+      },
+    },
+  },
+  async run() {
+    const products = await this.squarespace.getProducts();
```
Comment (feyzullah) [inline components/squarespace/sources/new-update-product/new-update-product.mjs:20]: Same comment as in `new-create-product`

File: components/squarespace/sources/new-update-transaction/new-update-transaction.mjs [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    squarespace,
+    timer: {
+      type: "$.interface.timer",
+      default: {
+        intervalSeconds: 60 * 5,
+      },
+    },
+  },
+  async run() {
+    const transactions = await this.squarespace.getTransactions();
```
Comment (feyzullah) [inline components/squarespace/sources/new-update-transaction/new-update-transaction.mjs:20]: Same comment as in `new-create-product`

**Review by feyzullah — COMMENTED** [review]
Hi @lcaresia I added some comments and questions. Can you please check?

File: components/squarespace/actions/create-product/create-product.mjs (hunk 1 above)
Comment (lcaresia) [reply components/squarespace/actions/create-product/create-product.mjs:21]: I chose to use `string` instead of `object` or `string[]` because this types are hard to parse and validate.

File: components/squarespace/sources/new-create-product/new-create-product.mjs (hunk 3 above)
Comment (lcaresia) [reply components/squarespace/sources/new-create-product/new-create-product.mjs:20]: Maybe I can add a filter, I will take a look.

File: components/squarespace/sources/new-create-transaction/new-create-transaction.mjs (hunk 4 above)
Comment (lcaresia) [reply components/squarespace/sources/new-create-transaction/new-create-transaction.mjs:20]: Just "new" transctions, the polling way first return all transactions, to after return just new.

**Review by jcortes — CHANGES_REQUESTED** [review]
Hi @lcaresia I've found only one issue in your PR. thanks!

**Review by jcortes — APPROVED** [review]

**Comment by vunguyenhung** [comment]
Hello everyone, I have tested this PR and there're some test cases failed or needed improvement. 

Please check the test report below for more information
[SquareSpace_2944.pdf](https://github.com/PipedreamHQ/pipedream/files/8941572/SquareSpace_2944.pdf)

**Comment by lcaresia** [comment]
@vunguyenhung I made a update, you can test again!

**Comment by vunguyenhung** [comment]
Hello @lcaresia, I have re-tested this PR and there's still a test cases failed.

Please check the test report below for more information
[Squarespace_3038_2010.pdf](https://github.com/PipedreamHQ/pipedream/files/8949132/Squarespace_3038_2010.pdf)

**Comment by lcaresia** [comment]
@vunguyenhung 

I think the problem is here, you are using double brackets

image

**Comment by vunguyenhung** [comment]
Hi everyone, all test cases are passed! Ready for release! 

Test report
[Squarespace_3038_2003.pdf](https://github.com/PipedreamHQ/pipedream/files/8967402/Squarespace_3038_2003.pdf)
